akka-docs/rst/general/handling-blocking-operations-in-akka-http-routes.rst

```diff
@@ -0,0 +1,167 @@
+Using ``context.dispatcher`` as the dispatcher on which the blocking Future
+executes, can be a problem. The same dispatcher is used by the routing
+infrastructure to actually handle the incoming requests. If all of the available
+threads are blocked, the routing infrastructure will end up starving. Therefore,
+routing infrastructure should not be blocked. Instead, a dedicated dispatcher
+for blocking operations should be used.
+
+Blocking APIs should also be avoided if possible. If it is not possible to avoid
+blocking APIs, then following solution explains how to handle blocking
+operations properly.
+
+
+Solution
+--------
+Following is the investigation of three pieces of code that explains the
+dispatcher impact and the performance of the app.
+
+The thread color state is represented as follows:-
+
+* Turquoise - Sleeping state
+* Orange - Waiting state
+* Green - Runnable state
+
+**1. Dispatcher behaviour on bad code**::
+
+  //Bad(due to blocking in Future):
+  implicit val defaultDispatcher = system.dispatcher
+  val routes: Route = post {
+  complete {
+    Future {                               // uses defaultDispatcher
+      Thread.sleep(5000)                     // will block on default dispatcher,
+      System.currentTimeMillis().toString    // Starving the routing infra
+      }
+    }
+  }
+
+Here the app is exposed to load of continous GET requests and large number
+of akka.actor.default-dispatcher threads are handling requests. The orange
+portion of the thread shows that they are idle.
+
+  .. image:: DispatcherBehaviourOnBadCode.png
+
+After some time, the app is exposed to the load of requesting POST requests,
+which will block these threads. For example "default-akka.default-dispatcher2,3,4"
+are going into the blocking state, after being idle before. It can be observed
+that the number of new threads increase, "default-akka.actor.default-dispatcher 18,
+19,20,..." however they go to sleep state immediately, thus wasting the
+resources.
+
+The number of such new threads depend on the default dispatcher configuration,
+but likely will not exceed 50. Since many POST requests are done, the entire
+thread pool is starved. The blocking operations dominate such that the routing
+infra has no thread available to handle the other requests.
+
+
+**2. Dispatcher behaviour good structured code/dispatcher**
+
+In ``application.conf``, the dispatcher dedicated for blocking behaviour should
+be configured as follows::
+
+  my-blocking-dispatcher {
+    type = Dispatcher
+    executor = "thread-pool-executor"
+    thread-pool-executer {
+      // in Akka previous to 2.4.2:
+      core-pool-size-min = 16
+      core-pool-size-max = 16
+      max-pool-size-min = 16
+      max-pool-size-max = 16
+      // or in Akka 2.4.2+
+      fixed-pool-size = 16
+    }
+      throughput = 100
+  }
+
+There are many dispatcher options available which can be found here
+`Akka Dispatchers <http://doc.akka.io/docs/akka/snapshot/scala/dispatchers.html>`_.
```

Comment:
This should be a :ref: link to the dispatchers chapter rather than an external link to the snapshot version of the docs